SageLink updates

joedakroub opened the pull request:
## Description
This PR updates the Sage Link rails implementation:
1. Uses `link_to` under-the-hood to generate links
2. Now has a `component.content` slot for adding custom/nested content
3. General refactoring and cleanup

## Screenshots
|  Before  |  After  |
|--------|--------|
|<img width="1792" alt="Before" src="https://user-images.githubusercontent.com/7331038/130256867-6cc8544e-5811-4ddb-adbe-11ad422dfe63.png">|<img width="1792" alt="After" src="https://user-images.githubusercontent.com/7331038/130281148-400cf710-7ff0-450d-8785-499988b4da36.png">|

## Testing in `sage-lib`
1. Visit http://0.0.0.0:4000/pages/component/link and ensure that the screen matches the `After` screenshot
2. Ensure all links are rendering properly in the DOM

## Testing in `kajabi-products`
1. (**HIGH**) Any use of the `SageLink` Rails component will be affected.
   - [ ] Account Password - /admin/settings/password
   - [ ] Contact Info - /admin/settings/contact_info
   - [ ] Account Users - /admin/settings/users

joedakroub commented on docs/app/views/examples/components/link/_preview.html.erb:
```diff
@@ -27,6 +27,14 @@
+    help_link: false,
+    show_label: true,
+  } do %>
+    <%= tag.span "Internal link with #{tag.strong "custom content"}".html_safe, class: "custom-content" %>
```
Example of providing custom/nested content.

joedakroub commented on docs/lib/sage_rails/app/views/sage_components/_sage_link.html.erb:
```diff
@@ -1,85 +1,38 @@
+html_attributes = {}
+  .merge(component.attributes || {})
+  .merge(component.html_attributes || {})
+label = component.content
```
CSS classes get stored in an array so that we can easily add classes bases on conditions.

`attributes` and `html_attributes` get merged into one hash that can be added to based on conditions.

philschanely replied:
It might be nice to take this chance to trim to using `html_attributes` instead of `attributes`. We have a ticket out about this effort more broadly, but IMO its a bug in our system that we allow both on some components. I def appreciate merging them here! Do you know if/how this method handles duplicate entries for a given key?

See #605

joedakroub replied:
I am definitely open to this. I just checked the app and the usage of `attributes` on SageLink is about 10 instances.

joedakroub commented on docs/lib/sage_rails/app/views/sage_components/_sage_link.html.erb:
```diff
@@ -1,85 +1,38 @@
+  css_classes.append("sage-link--no-launch")
+  html_attributes[:rel] = "noopener noreferrer"
+  html_attributes[:target] = "_blank"
+end
```
Here we added classes and set properties of `html_attributes` to be used later in the `link_to` call.

joedakroub commented on docs/lib/sage_rails/app/views/sage_components/_sage_link.html.erb:
```diff
@@ -1,85 +1,38 @@
+  .push(component.generated_css_classes)
+  .join(" ")
+  .squish
 %>
```
Add the `generated_css_classes` last, join the array into a space-separated string and trim.

joedakroub commented on docs/lib/sage_rails/app/views/sage_components/_sage_link.html.erb:
```diff
@@ -1,85 +1,38 @@
-    <%= label %>
-  </a>
-<% end %>
+<%= link_to label, component.url, html_attributes %>
```
Call `link_to` with the label, url and attributes.

joedakroub commented on docs/lib/sage_rails/app/views/sage_components/_sage_link.html.erb:
```diff
@@ -1,85 +1,38 @@
+
+if component.label.present?
+  label = if component.truncate.present?
+    tag.span component.label, class: "t-sage--truncate"
```
Using the `tag` helper to generate a `span` element.

pixelflips commented on docs/lib/sage_rails/app/views/sage_components/_sage_link.html.erb:
```diff
@@ -1,85 +1,38 @@
+  css_classes.append("sage-link--help-icon-only")
+  label = tag.span label, class: "visually-hidden"
+elsif !component.external && component.launch
+  css_classes.append("sage-link--launch")
```
Not sure if this update is actually the cause, but I thought I remembered our launch links had a `launch` icon? I see the styles in the _link.scss file, but it doesn't seem to render. 

I see it no longer appears on the live docs site either. If we no longer need the icon, it might be worth it to remove the styles.

joedakroub replied:
Good catch Phillip! I'll take a look into that. Strange that the live docs are displaying this as well. I know we used to have launch icons here.

joedakroub replied:
Looks like the launch links need a `span.t-sage--truncate` wrapper. I added logic in to address this. Launch icon is back. 👍🏼

anechol commented:
Very awesome update!!

pixelflips reviewed (approved):
LGTM! 👍🏼

monkeypox8 commented on docs/lib/sage_rails/app/views/sage_components/_sage_link.html.erb:
```diff
@@ -1,85 +1,39 @@
+  html_attributes[:target] = "_blank"
+elsif component.external && !component.launch
+  css_classes.append("sage-link--no-launch")
+  html_attributes[:rel] = "noopener noreferrer"
```
This brings up something that I've been pondering for a while now after seeing similar use thoroughout Sage: should we continue to apply both of these attributes for "launch" links in components?

If I understand the usage of the referral header correctly, there would seem to be cases where we'd actually want it to be accounted for. Zendesk help links are a good example. But if we do feel that we would like to keep the existing `noreferrer` behavior, we should be able to drop the use of `noopener`, since that keyword is implied with `noreferrer`.

joedakroub replied:
I am able to replicate this error as well. Looking into it now. I removed `noopener` from both; good point. 👍🏼

joedakroub replied:
I was able to correct the broken views mentioned above. The error was due to the use of the `tag` helper. I switched to using the `content_tag` helper and the issues were fixed.

monkeypox8 reviewed (commented):
I'm seeing a couple of broken views running the bridge, and judging by their error lists, I suspect it may be related to the `html_attributes` and `attributes` merge in the component. Anyone else able to verify?

<details>
<summary>Products list</summary>
<img src="https://user-images.githubusercontent.com/816579/130302659-5a669451-757f-4d4b-b778-afbd7b99d108.png" alt="">
</details>

<details>
<summary>Offer edit</summary>
<img src="https://user-images.githubusercontent.com/816579/130302719-4cc41a52-ad83-4ca8-b339-51d284227e28.png" alt="">
</details>

philschanely commented:
Related: #688

philschanely reviewed (approved):
Fabulous work here. Excited about the outcome.

joedakroub commented:
@philschanely The `merge` function will overwrite the props of the previous hash with the next. So if `attributes` contains an `href=123` and `html_attributes` contains `href=456`, it will take the last value.

joedakroub commented:
@philschanely I removed the `attributes` prop and updated the example to use `html_attributes`. Will create a PR in `kajabi-products` to address the `SageLink` instances. 👍🏼

monkeypox8 reviewed (approved):
👍🏼 LGTM! The offer edit page (and other views with the shared automations partial) are currently failing on this branch due to the use of `attributes`, but it's working great locally with the related patch + bridge. Nice work, this will be a valuable addition!
